### ISSUE TITLE
New package proposal

### ISSUE THREAD

**Issue by ctron** [eclipse/packages#274]
I would like to propose a new package, consisting of the following stack:

* Device/Sensor
  * Drogue IoT Device
* Gateway
  * Eclipse ioFog (as the edge platform)
  * Eclipse Kura (as the IoT workload on the edge)
* Cloud
  * Drogue IoT Cloud (as the IoT connectivity layer)
  * Apache Kafka (as the event streaming solution)
  * Knative Eventing (as the event orchestration)
  * Eclipse Ditto (as digital twin component)
  * Eclipse Streamsheets (as data visualization solution)

The goal of the package is to show an end-to-end (device-to-cloud) architecture with open source components, focusing on the use-case of telemetry data acquisition.

We are willing to drive this project, but would hope that the respective projects would be supportive in this effort.

We also did some initial tests, to check out if this is technically feasible. And so far, this looks good. There are a few tweaks we had to do on the different components, but we hope that these tweaks will land in the respective projects, rather than the IoT Packages project.

Regarding the device and edge layer, we plan two approaches:

1.) Use a real microcontroller (STM32 or nRF) and and a Raspberry Pi (Zero W or better)
2.) Emulate the micrcontroller using WebAssembly in the browser and the edge node using a virtual machine

As a first step, I am asking for feedback on this proposal. If you like please indicate that you are interested in supporting this effort.

**Comment by thjaeckle** [comment]
Hi @ctron 
Is this something still intended to be executed on a local machine? Seems for me like a "cloud only" setup.

I could support in the Ditto part of such a package and would suggest to push data into Streamsheets utilizing the "Digital Twin" of Ditto (e.g. via HTTP push, doing conversion to a cloud events inside of a custom Ditto payload mapper) in order to work on a normalized data structure (the Ditto "Thing" and Ditto Protocol events).

**Comment by ctron** [comment]
> Is this something still intended to be executed on a local machine? Seems for me like a "cloud only" setup.

Yes, it should be possible to test this on your local machine. It still requires "Kubernetes" and "Helm" as intended for the IoT Packages.

You might need a bit of RAM to run this. But aside from Kafka, Keycloak and Ditto, the different components don't need much. Not adding Ditto, we can fire this up on a GitHub actions runner, which has 7.5GiB RAM available.

What might be a bit tricky is the "non-cloud" part of the edge node and the microcontroller for the sensor. Thus the idea to emulate this.

> I could support in the Ditto part of such a package and would suggest to push data into Streamsheets utilizing the "Digital Twin" of Ditto (e.g. via HTTP push, doing conversion to a cloud events inside of a custom Ditto payload mapper) in order to work on a normalized data structure (the Ditto "Thing" and Ditto Protocol events).

That would be awesome, I am happy to count you in. My idea was to have some Ditto-normalized payload between Ditto and Streamsheets. And Ditto would receive the data from Kafka using the Knative eventing (KafkaSource) and the Ditto Cloud Events endpoint, with a "serverless" function in the middle. But we can definitely come up with alternative ideas, thus the call for feedback :grinning:

**Comment by thjaeckle** [comment]
The upcoming Ditto 2.1.0 can directly consume Kafka messages and also can apply Javascript based custom mapping.

Using knative eventing would probably be the more modern/serverless way, however when using Ditto as stateful server in that setup, one may also benefit from its features 😉

**Comment by ctron** [comment]
> The upcoming Ditto 2.1.0 can directly consume Kafka messages and also can apply Javascript based custom mapping.
> 
> Using knative eventing would probably be the more modern/serverless way, however when using Ditto as stateful server in that setup, one may also benefit from its features wink

That is an interesting feature. Definitely worth to explore. The idea behind using Knative was to re-use some existing ecosystem, which provides exactly that (and also not talk people into using JavaScript :grin:). Maybe we can even leverage the Knative eventing registry for automatic payload conversion.

Maybe let's start out with what we have, and then do an update (or addon) later on.

**Comment by ctron** [comment]
As discussed on Gitter, we would like to move ahead with the idea of using the JavaScript functionality. That should make the deployment a bit more light weight, and we can also add Knative on top.

**Comment by MMaiero** [comment]
Hi Jens,
thanks for bringing this up. Please count us for this effort.

Matteo
